### PR TITLE
Quick and probably nasty kludge to introduce increased entropy method…

### PR DESCRIPTION
**Pull request by Oneiroi**
… for the sessionStoreKey

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -80,8 +79,7 @@ func main() {
 
 	loginError = WebAuthnError{Message: "Unable to login"}
 	registrationError = WebAuthnError{Message: "Error during registration"}
-
-	rand.Seed(time.Now().UnixNano())
+	util.RandInit()
```
Comment (Oneiroi) [inline main.go]: on reflection this may be redundant now and can be removed

File: util/util.go [hunk 2]
```diff
@@ -83,11 +87,35 @@ func SaveWebauthnSession(session *sessions.Session, key string, sessionData *web
 	return nil
 }
 
+func RandInit() {
```
Comment (Oneiroi) [inline util/util.go]: this method may be able to be cleaned up, however leaving this in place _incase_ I missed anywhere.

File: util/util.go [hunk 3]
```diff
@@ -83,11 +87,35 @@ func SaveWebauthnSession(session *sessions.Session, key string, sessionData *web
...
+	_, err := crypto_rand.Read(challenge)
+	if err != nil {
+		panic("failed to seed challenge from crypto/rand cryptographically secure function")
+	}
+	r := base64.RawURLEncoding.EncodeToString(challenge)
+	return r
+}
+
 // Generate a random string of alpha characters of length n
 func RandStringBytesRmndr(n int) []byte {
```
Comment (Oneiroi) [inline util/util.go]: As with `RandInit()` this _may_ be able to be removed.

**Comment by Oneiroi** [comment]
*this has not been functionally tested at this moment in time* please lmk if there is some areas I have missed, thanks!

**Comment by Oneiroi** [comment]
improvement for #2 to raise entropy shannon scale to ~6.02 for `sessionStoreKey`

**Comment by Oneiroi** [comment]
Yeah I've a _lot_ to cleanup here, bare with me 😓

**Comment by Oneiroi** [comment]
_alright_ @pdavies011010 please t.a.l, I suspect this is probably horrible, however this should be tested and _if_ functional can be merged :+1:

File: util/util.go [hunk 4]
```diff
@@ -95,6 +97,16 @@ func RandInit() {
...
+	//call on the import DUO method
+	challenge, err := protocol.CreateChallenge()
+	if err != nil {
+		panic("Failed to generate cryographically secure challenge")
+	}
+	return base64.RawURLEncoding.EncodeToString(challenge)
+}
+
 // Generate a random string of alpha characters of length n
 func RandStringBytesRmndr(n int) []byte {
```
Comment (pdavies011010) [inline util/util.go]: Hey @Oneiroi Yep looks like you can remove RandInit and RandStringBytesRmndr

**Review by pdavies011010 — CHANGES_REQUESTED** [review]
Looks good @Oneiroi , if you just want to remove those two functions we'll get this merged in and tested

File: util/util.go (hunk 4 above)
Comment (Oneiroi) [reply util/util.go]: Removed within #[f7f4776](https://github.com/Quiq/webauthn_proxy/pull/3/commits/f7f47768feb76d64a577beefdc78b2e580702616)

File: util/util.go (hunk 3 above)
Comment (Oneiroi) [reply util/util.go]: Removed within #[f7f4776](https://github.com/Quiq/webauthn_proxy/pull/3/commits/f7f47768feb76d64a577beefdc78b2e580702616)

File: util/util.go (hunk 2 above)
Comment (Oneiroi) [reply util/util.go]: Removed within #[f7f4776](https://github.com/Quiq/webauthn_proxy/pull/3/commits/f7f47768feb76d64a577beefdc78b2e580702616)

File: main.go (hunk 1 above)
Comment (Oneiroi) [reply main.go]: this was removed in b30556a

File: util/util.go (hunk 4 above)
Comment (Oneiroi) [reply util/util.go]: over to you @pdavies011010 🙇

**Review by pdavies011010 — APPROVED** [review]

**Pull request merged**